### PR TITLE
Rule Balls in Heatmap

### LINKED ISSUE #514: Use `SubState` in ball search
If information about possible ball positions can be extracted from the current game state, we should use this information for ball search.
Some ideas/examples of where to search when:
- corners during `SubState::CornerKick`
- goalkick positions during `SubState::GoalKick`
- field boundaries close to the last known ball position during `SubState::KickIn`
- penalty spot(s) during `SubState::PenaltyKick`

### LINKED ISSUE #994: Ruleballs in Heatmap
Add search hypotheses for corner kick, ...

### PR DESCRIPTION
**Pull request by knoellle**
## Why? What?

Adds "heat" to the ball search heat map in positions where a ball might be according to the rules.
Currently covers kick off, penalty kicks, goal kicks and corner kicks.
Not handled are kick-in and pushing free kick.
Kick-in because it can happen on any point on the side lines.
Pushing free kick does **not** need special handling because the ball doesn't (or isn't supposed to) move.

N.b. I haven't tested this with a game controller yet, so I'm not sure if the conditions are quite correct.

Fixes #994 
Fixes #514

## Ideas for Next Iterations (Not This PR)

Handle kick-in somehow, maybe by projecting the existing heat to the nearest side line.

## How to Test

Check heatmap in the relevant situations.

### CODE REVIEW

File: crates/control/src/search_suggestor.rs [hunk 1]
```diff
@@ -151,3 +161,45 @@ impl IndexMut<Point2<Field>> for Heatmap {
...
+            vec![
+                field_dimensions.corner(kick_half, Side::Left),
+                field_dimensions.corner(kick_half, Side::Right),
+            ]
+        }
+        (PrimaryState::Playing, Some(SubState::GoalKick)) => {
+            let kick_half = kicking_team_half;
+            vec![
+                field_dimensions.corner(kick_half, Side::Left),
+                field_dimensions.corner(kick_half, Side::Right),
```
Comment (MaikRe) [inline crates/control/src/search_suggestor.rs:191]:
```suggestion
        (PrimaryState::Playing, Some(SubState::GoalKick)) => {
            let kick_half = kicking_team_half;
            vec![
                field_dimensions.goal_box_corner(kick_half, Side::Left),
                field_dimensions.goal_box_corner(kick_half, Side::Right),
```

File: crates/control/src/search_suggestor.rs [hunk 2]
```diff
@@ -151,3 +161,45 @@ impl IndexMut<Point2<Field>> for Heatmap {
...
+fn get_rule_hypotheses(
+    primary_state: PrimaryState,
+    filtered_game_controller_state: FilteredGameControllerState,
+    field_dimensions: FieldDimensions,
+) -> Vec<Point2<Field>> {
+    let kicking_team_half =
+        conservative_kicking_team_half(filtered_game_controller_state.kicking_team);
+
+    match (primary_state, filtered_game_controller_state.sub_state) {
+        (PrimaryState::Ready, Some(SubState::PenaltyKick)) => {
```
Comment (MaikRe) [inline crates/control/src/search_suggestor.rs:174]: Should this also include the primary state Set?

File: crates/control/src/search_suggestor.rs [hunk 3]
```diff
@@ -151,3 +161,45 @@ impl IndexMut<Point2<Field>> for Heatmap {
...
+            let kick_half = kicking_team_half.mirror();
+            vec![field_dimensions.penalty_spot(kick_half)]
+        }
+        // Kick-off
+        (PrimaryState::Ready, None) => vec![field_dimensions.center()],
```
Comment (MaikRe) [inline crates/control/src/search_suggestor.rs:179]: This both seems like it overlaps/duplicates a bit with the ```rule_ball``` in the ```ball_state_composer```. 
I would suggest maybe using the rule ball as an input for the search suggestor, what do you think?

**Review by MaikRe — CHANGES_REQUESTED** [review]

File: crates/control/src/search_suggestor.rs (hunk 3 above)
Comment (MaikRe) [reply crates/control/src/search_suggestor.rs:179]: Fair enough

File: crates/control/src/search_suggestor.rs (hunk 2 above)
Comment (knoellle) [reply crates/control/src/search_suggestor.rs:174]: I'm not sure about the exact ball search mechanics, but during set no one should be doing ball search anyways.

@JonasHolz do you think it makes sense to apply rule ball positions during set also?

File: crates/control/src/search_suggestor.rs (hunk 1 above)
Comment (knoellle) [reply crates/control/src/search_suggestor.rs]: Done

**Comment by MaikRe** [comment]
Does this also fix #514?

File: crates/control/src/search_suggestor.rs (hunk 2 above)
Comment (MaikRe) [reply crates/control/src/search_suggestor.rs:174]: macht zu mir sinn

**Review by MaikRe — APPROVED** [review]
Arbeitet wie in der Werbung

**Pull request merged**